### PR TITLE
[SPARK-30574][DOC] Document GROUP BY Clause of SELECT statement in SQL Reference.

### PR DESCRIPTION
**Pull request by dilipbiswal**
### What changes were proposed in this pull request?
Document GROUP BY clause of SELECT statement in SQL Reference Guide.

### Why are the changes needed?
Currently Spark lacks documentation on the supported SQL constructs causing
confusion among users who sometimes have to look at the code to understand the
usage. This is aimed at addressing this issue.

### Does this PR introduce any user-facing change?
Yes. 

**Before:**
There was no documentation for this.

**After.**
Screen Shot 2020-01-19 at 5 11 12 PM
Screen Shot 2020-01-19 at 5 11 32 PM
Screen Shot 2020-01-19 at 5 11 49 PM
Screen Shot 2020-01-19 at 5 12 05 PM
Screen Shot 2020-01-19 at 5 12 31 PM
Screen Shot 2020-01-19 at 5 26 38 PM

### How was this patch tested?
Tested using jykyll build --serve

### CODE REVIEW

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 1]
```diff
@@ -18,5 +18,206 @@ license: |
...
+    Groups the rows for each subset of the expressions specified in the grouping sets. For example, 
+    <code>GROUP BY GROUPING SETS (warehouse, product)</code> is semantically equivalent
+    to union of results of <code>GROUP BY warehouse</code> and <code>GROUP BY product</code>. This clause
+    is a short-hand to using an <code>UNION ALL</code> where each leg of the <code>UNION ALL</code> 
+    operator performs aggregation of subset of the columns specified in the <code>GROUPING SETS</code> clause.
+  </dd>
+  <dt><code><em>grouping_sets</em></code></dt>
+  <dd>
+    Specifies one of more groupings based on which the <code>GROUP BY</code> clause performs aggregations. A grouping
+    set is specified by a list of comma separated expressions in parenthses.<br><br>
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-groupby.md]: ```parenthses``` typo

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 2]
```diff
@@ -18,5 +18,206 @@ license: |
...
+    <code>GROUP BY</code> clause. For example, <code>GROUP BY warehouse, product  WITH CUBE</code> is equivalent 
+    to GROUP BY warehouse, product GROUPING SETS ((warehouse, product), (warehouse), (product), ()).
+    The N elements of a <code>CUBE</code> specification results in 2^N <code>GROUPING SETS</code>.
+  </dd>
+</dl>
+
+### Examples
+{% highlight sql %}
+CREATE TABLE dealer (id INT, city STRING, car_model STRING, quantity INT);
+INSERT INTO dealer  VALUES (100, 'Fremont', 'Honda Civic', 10),
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-groupby.md]: There is an extra space between ```dealer``` and   ```VALUES```

**Comment by huaxingao** [comment]
```displayTitle: GROUPBY Clause``` -> ```displayTitle: GROUP BY Clause```?

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 3]
```diff
@@ -18,5 +18,206 @@ license: |
...
+{% endhighlight %}
+
+### Parameters
+<dl>
+  <dt><code><em>GROUPING SETS</em></code></dt>
+  <dd>
+    Groups the rows for each subset of the expressions specified in the grouping sets. For example, 
+    <code>GROUP BY GROUPING SETS (warehouse, product)</code> is semantically equivalent
+    to union of results of <code>GROUP BY warehouse</code> and <code>GROUP BY product</code>. This clause
+    is a short-hand to using an <code>UNION ALL</code> where each leg of the <code>UNION ALL</code>
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-groupby.md]: ```This clause is a short-hand to using an <code>UNION ALL</code>``` 
-> ```This clause is a shorthand for a <code>UNION ALL</code>``` ?

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 4]
```diff
@@ -18,5 +18,206 @@ license: |
...
+    Groups the rows for each subset of the expressions specified in the grouping sets. For example, 
+    <code>GROUP BY GROUPING SETS (warehouse, product)</code> is semantically equivalent
+    to union of results of <code>GROUP BY warehouse</code> and <code>GROUP BY product</code>. This clause
+    is a shorthand for a <code>UNION ALL</code> where each leg of the <code>UNION ALL</code> 
+    operator performs aggregation of subset of the columns specified in the <code>GROUPING SETS</code> clause.
+  </dd>
+  <dt><code><em>grouping_sets</em></code></dt>
+  <dd>
+    Specifies one of more groupings based on which the <code>GROUP BY</code> clause performs aggregations. A grouping
+    set is specified by a list of comma separated expressions in parenthesis.<br><br>
```
Comment (srowen) [inline docs/sql-ref-syntax-qry-select-groupby.md]: comma-separated
parentheses

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 5]
```diff
@@ -18,5 +18,206 @@ license: |
...
+    <b>Syntax:</b>
+      <code>
+        (() | (expression [ , ...]))
+      </code>
+  </dd>
+  <dt><code><em>grouping_expression</em></code></dt>
+  <dd>
+    Specifies the crieteria based on which the rows are grouped together. A grouping expression may be a column alias,
```
Comment (srowen) [inline docs/sql-ref-syntax-qry-select-groupby.md]: critieria
Somewhere above or here do you want to say that the grouping is done on the _values_ that result from the expression? calling them criteria suggests they are logical expressions, perhaps.

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 6]
```diff
@@ -18,5 +18,206 @@ license: |
...
+{% endhighlight %}
+
+### Parameters
+<dl>
+  <dt><code><em>GROUPING SETS</em></code></dt>
+  <dd>
+    Groups the rows for each subset of the expressions specified in the grouping sets. For example, 
+    <code>GROUP BY GROUPING SETS (warehouse, product)</code> is semantically equivalent
+    to union of results of <code>GROUP BY warehouse</code> and <code>GROUP BY product</code>. This clause
+    is a shorthand for a <code>UNION ALL</code> where each leg of the <code>UNION ALL</code>
```
Comment (srowen) [inline docs/sql-ref-syntax-qry-select-groupby.md]: I'd say "is shorthand", but either way be consistent as "short-hand" appears below

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 7]
```diff
@@ -18,5 +18,206 @@ license: |
...
+    Specifies multiple levels of aggregations in a single statement. This clause is used to compute aggregations 
+    based on multiple grouping sets. <code>ROLLUP</code> is a short-hand for <code>GROUPING SETS</code>. For example,
+    GROUP BY warehouse, product  WITH ROLLUP is equivalent to GROUP BY warehouse, product GROUPING SETS ((warehouse, product), (warehouse), ()).
+    The N elements of a <code>ROLLUP</code> specification results in N+1 <code>GROUPING SETS</code>.
+  </dd>
+  <dt><code><em>CUBE</em></code></dt>
+  <dd>
+    <code>CUBE</code> clause is used to perform aggregations based on combination of grouping columns specified in the 
+    <code>GROUP BY</code> clause. For example, <code>GROUP BY warehouse, product  WITH CUBE</code> is equivalent 
+    to GROUP BY warehouse, product GROUPING SETS ((warehouse, product), (warehouse), (product), ()).
```
Comment (srowen) [inline docs/sql-ref-syntax-qry-select-groupby.md]: Code-font this expression?

**Review by srowen — CHANGES_REQUESTED** [review]

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**